File: weyl-core/src/main/java/io/quartic/weyl/core/source/SourceManager.java
package io.quartic.weyl.core.source;

import io.quartic.catalogue.CatalogueEvent;
import io.quartic.catalogue.api.model.DatasetConfig;
import io.quartic.catalogue.api.model.DatasetId;
import io.quartic.catalogue.api.model.DatasetMetadata;
import io.quartic.weyl.core.model.LayerId;
import io.quartic.weyl.core.model.LayerMetadata;
import io.quartic.weyl.core.model.LayerPopulator;
import io.quartic.weyl.core.model.LayerSpec;
import io.quartic.weyl.core.model.LayerUpdate;
import io.quartic.weyl.core.model.MapDatasetExtension;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import rx.Observable;
import rx.Scheduler;
import rx.observables.GroupedObservable;

import java.util.Optional;
import java.util.function.Function;

import static io.quartic.catalogue.CatalogueEvent.Type.CREATE;
import static io.quartic.catalogue.CatalogueEvent.Type.DELETE;
import static java.lang.String.format;
import static rx.Observable.empty;
import static rx.Observable.just;

public class SourceManager {
    private static final Logger LOG = LoggerFactory.getLogger(SourceManager.class);
    private final Observable<CatalogueEvent> catalogueEvents;
    private final Function<DatasetConfig, Optional<Source>> sourceFactory;
    private final Scheduler scheduler;
    private final ExtensionCodec extensionCodec;
    private Observable<LayerPopulator> layerPopulators = null;  // TODO - eliminate grossness to emulate lazy evaluation

    public SourceManager(
            Observable<CatalogueEvent> catalogueEvents,
            Function<DatasetConfig, Optional<Source>> sourceFactory,
            Scheduler scheduler
    ) {
        this(catalogueEvents, sourceFactory, scheduler, new ExtensionCodec());
    }

    public SourceManager(
            Observable<CatalogueEvent> catalogueEvents,
            Function<DatasetConfig, Optional<Source>> sourceFactory,
            Scheduler scheduler,
            ExtensionCodec extensionCodec
    ) {
        this.catalogueEvents = catalogueEvents;
        this.sourceFactory = sourceFactory;
        this.scheduler = scheduler;
        this.extensionCodec = extensionCodec;
    }

    public Observable<LayerPopulator> layerPopulators() {
        if (layerPopulators == null) {
            layerPopulators = catalogueEvents
                    .groupBy(CatalogueEvent::getId)
                    .flatMap(this::processEventsForId)
                    .share();
        }
        return layerPopulators;
    }

    private Observable<LayerPopulator> processEventsForId(GroupedObservable<DatasetId, CatalogueEvent> group) {
        final Observable<CatalogueEvent> events = group.cache();    // Because groupBy can only have one subscriber per group
        final Observable<LayerPopulator> populator = events
                .filter(e -> e.getType() == CREATE)
                .flatMap(event -> createSource(event.getId(), event.getConfig())
                        .map(source -> createPopulator(
                                event.getId(),
                                event.getConfig(),
                                sourceUntil(source, deletionEventFrom(events))))
                );

        LOG.info("[{}] Finished constructing populator", group.getKey());
        return populator;
    }

    private LayerPopulator createPopulator(DatasetId id, DatasetConfig config, Source source) {
        final String name = config.getMetadata().getName();
        final MapDatasetExtension extension = extensionCodec.decode(name, config.getExtensions());

        LOG.info(format("[%s] Created layer", name));

        return LayerPopulator.withoutDependencies(
                new LayerSpec(
                        new LayerId(id.getUid()),
                        datasetMetadataFrom(config.getMetadata()),
                        extension.getViewType().getLayerView(),
                        extension.getStaticSchema(),
                        source.indexable()
                ),
                source.observable().subscribeOn(scheduler)     // TODO: the scheduler should be chosen by the specific source;
        );
    }

    private Observable<Source> createSource(DatasetId id, DatasetConfig config) {
        try {
            Optional<Source> source = sourceFactory.apply(config);
            if (!source.isPresent()) {
                LOG.error(format("[%s] Unhandled config : %s", id, config.getLocator()));
                return empty();
            }

            return just(source.get());
        } catch (Exception e) {
            LOG.error(format("[%s] Error creating layer for dataset", id), e);
            return empty();
        }
    }

    // This mechanism causes the observable to complete, as well as unsubscription from the upstream
    private Source sourceUntil(Source source, Observable<?> until) {
        return new Source() {
            @Override
            public Observable<LayerUpdate> observable() {
                return source.observable().takeUntil(until);
            }

            @Override
            public boolean indexable() {
                return source.indexable();
            }
        };
    }

    private Observable<CatalogueEvent> deletionEventFrom(Observable<CatalogueEvent> events) {
        return events
                .filter(e -> e.getType() == DELETE)
                .doOnNext(e -> LOG.info(format("[%s] Deleted layer", e.getConfig().getMetadata().getName())));
    }

    // TODO: do we really need LayerMetadata to be distinct from DatasetMetadata?
    private LayerMetadata datasetMetadataFrom(DatasetMetadata metadata) {
        return new LayerMetadata(
                metadata.getName(),
                metadata.getDescription(),
                metadata.getAttribution(),
                metadata.getRegistered()    // Should always be non-null
        );
    }
}

File: weyl-core/src/main/java/io/quartic/weyl/core/source/SourceManager.kt
package io.quartic.weyl.core.source

import io.quartic.catalogue.CatalogueEvent
import io.quartic.catalogue.CatalogueEvent.Type.CREATE
import io.quartic.catalogue.CatalogueEvent.Type.DELETE
import io.quartic.catalogue.api.model.DatasetConfig
import io.quartic.catalogue.api.model.DatasetCoordinates
import io.quartic.catalogue.api.model.DatasetMetadata
import io.quartic.common.logging.logger
import io.quartic.weyl.core.model.LayerId
import io.quartic.weyl.core.model.LayerMetadata
import io.quartic.weyl.core.model.LayerPopulator
import io.quartic.weyl.core.model.LayerSpec
import rx.Observable
import rx.Observable.empty
import rx.Observable.just
import rx.Scheduler
import rx.observables.GroupedObservable
import java.lang.String.format

class SourceManager @JvmOverloads constructor(
        private val catalogueEvents: Observable<CatalogueEvent>,
        private val sourceFactory: (DatasetConfig) -> Source?,
        private val scheduler: Scheduler,
        private val extensionCodec: ExtensionCodec = ExtensionCodec()
) {
    private val LOG by logger()

    val layerPopulators by lazy {
        catalogueEvents
                .groupBy { it.coords }
                .flatMap { this.processEventsForCoords(it) }
                .share()
    }

    private fun processEventsForCoords(group: GroupedObservable<DatasetCoordinates, CatalogueEvent>): Observable<LayerPopulator> {
        val events = group.cache()    // Because groupBy can only have one subscriber per group
        return events
                .filter { it.type === CREATE }
                .flatMap {
                    createSource(it.coords, it.config)
                            .map { source -> createPopulator(
                                    it.coords,
                                    it.config,
                                    sourceUntil(source, deletionEventFrom(events))
                            )}
                            .flatMap { s -> s }
                }
    }

    private fun createPopulator(coords: DatasetCoordinates, config: DatasetConfig, source: Source): Observable<LayerPopulator> {
        val name = config.metadata.name
        val extension = extensionCodec.decode(name, config.extensions)

        return if (extension == null) {
            empty()
        } else {
            LOG.info("[$name] Created layer")
            just(LayerPopulator.withoutDependencies(
                    LayerSpec(
                            LayerId(coords.toString()), // // TODO - deal with DatasetCoords properly
                            datasetMetadataFrom(config.metadata),
                            extension.viewType.layerView,
                            extension.staticSchema,
                            source.indexable()
                    ),
                    source.observable().subscribeOn(scheduler)     // TODO: the scheduler should be chosen by the specific source;
            ))
        }
    }

    private fun createSource(coords: DatasetCoordinates, config: DatasetConfig): Observable<Source> {
        try {
            val source = sourceFactory(config)
            if (source != null) {
                return just(source)
            }
            LOG.error("[$coords] Unhandled config : ${config.locator}")
        } catch (e: Exception) {
            LOG.error("[$coords] Error creating layer for dataset", e)
        }
        return empty()
    }

    // This mechanism causes the observable to complete, as well as unsubscription from the upstream
    private fun sourceUntil(source: Source, until: Observable<*>) = object : Source {
        override fun observable() = source.observable().takeUntil(until)
        override fun indexable() = source.indexable()
    }

    private fun deletionEventFrom(events: Observable<CatalogueEvent>) = events
            .filter { it.type === DELETE }
            .doOnNext { LOG.info("[${it.config.metadata.name}] Deleted layer") }

    // TODO: do we really need LayerMetadata to be distinct from DatasetMetadata?
    private fun datasetMetadataFrom(metadata: DatasetMetadata) = LayerMetadata(
            metadata.name,
            metadata.description,
            metadata.attribution,
            metadata.registered!!    // Should always be non-null
    )
}
